Refactor Game_Character classes 1 of 2

fmatthew5876 opened the pull request:
Depends on: https://github.com/EasyRPG/liblcf/pull/282

This PR refactors the `Game_Character` class heirarchy to use the `RPG::SaveMapEventBase` class heirarchy from liblcf.

The goals are:
- [x] Correct loading of fields from LSD
- [x] Correct saving of fields to LSD
- [x] Correct behavior when saving / loading between `RPG_RT` <-> `Player`.
- [x] Binary compatibility with `RPG_RT` even when data change appears not change behavior.
- [x] Devirtualize parts of `Game_Character` where not needed
- [x] Simplify and remove duplication of logic and state
- [x] Resolve #1499

fmatthew5876 commented on src/game_character.cpp:
```diff
@@ -851,12 +850,8 @@ bool Game_Character::IsInPosition(int x, int y) const {
-
 int Game_Character::GetOpacity() const {
-	return opacity;
+	return (8 - GetTransparency()) * 32 - 1;
```
@CherryDT Do you agree with this calculation?

Transparency is valued from 0 to 7, with 7 being almost completely transparent. This value can't go one more step to 8 to set fully transparent.

CherryDT replied:
Almost. It is a bit weirder. It is `255 - GetTransparency() * 30`. Yes, **30** and not 32.

Ghabry commented on src/game_character.cpp:
```diff
@@ -810,7 +809,7 @@ void Game_Character::CancelMoveRoute() {
 int Game_Character::GetTileId() const {
-	return tile_id;
+	return GetSpriteName().empty() ? GetSpriteIndex() : 0;
 }
```
> Simplify GetTileId() 
should be tested against this test case: https://github.com/EasyRPG/Player/pull/1179

Ghabry commented on src/game_vehicle.cpp:
```diff
@@ -296,7 +290,6 @@ void Game_Vehicle::Update() {
 			if (!IsDescending()) {
 				if (CanLand()) {
 					SetLayer(RPG::EventPage::Layers_same);
```
> Remove driving

Issue worth testing: https://github.com/EasyRPG/Player/issues/1349

Ghabry replied:
Doesn#t break the linked testcase and also not the Yume2kki music room.

fmatthew5876 commented:
Unless I find more issues to fix, I think this PR is nearly done. It's a lot of commits, so once the liblcf dependencies are in it could be broken up if needed.

Here are some save chunks I'm still unsure about:

* `map_id` - For some reason `RPG_RT` always sets `map_id = 0` for all map events. The map_id for the player and vehicles is set correctly. I'm hesitant to do this as someone could call GetMapId() on a game_event in the code expecting the right Id and getting surprised.
* `processed` - This chunk is used somehow by `RPG_RT` to resolve movement, particularly when you have 2 events that want to move and event A wants to take event B's spot. It looks like some kind of way to not double compute event movement. How do we handle this situation in Player? At some point we should maybe also use this flag to ensure our saved games don't break in mysterious ways between RPG_RT and Player.
* `route_through` - Move commands that set/unset through also set this flag. I set it for binary compatibility on LSD but I don't use it. Why does this exist? What edge case do we need to differentiate between `through` and `route_through`? If there is such a case, fixing in it Player would likely fix esoteric event code bugs in some game.

fmatthew5876 commented:
For `processed`, I decided to just write the flag so it looks like `RPG_RT`.

My observation is that `processed` will always be 1 in `RPG_RT` saves (except for vehicles you haven't seen yet). My assumption is that this is because you can never save at a time when events are resolving their moves and haven't finished processing.

Is that assumption correct?

carstene1ns commented:
jenkins: you know what to do, test this please :heart_eyes:

fmatthew5876 commented:
@Ghabry and @carstene1ns 

Would you like me to break this one up into separate PRs?

carstene1ns commented:
> game_character.h(1100): error C2589: '(': illegal token on right side of '::'

The weird build error happens, because windows header has own macros for min() and max().
Needs [this](https://github.com/EasyRPG/Player/blob/master/src/sdl2_ui.cpp#L29-L31) stuff [here](https://github.com/EasyRPG/Player/blob/master/src/filefinder.cpp#L31).

Ghabry commented:
> Would you like me to break this one up into separate PRs?

@fmatthew5876 
No, I think it's fine. Because is mostly code simplification because of the new inheritance feature in liblcf I think we can handle this.
Will tell you when I find anything that could be extracted.

fmatthew5876 commented on src/utils.h:
```diff
@@ -234,6 +234,17 @@ namespace Utils {
+	 * @return value clamped between min and max
+	 */
+	template <typename T>
+		T Clamp(T value, const T& minv, const T& maxv);
```
Fixed.

fmatthew5876 commented:
jenkins: test this please

Ghabry commented:
I'm requesting a split after the commit "Remove Game_Vehicle::driving" with the commit "Add compatibility hack for old Player save games" cherry-picked. (using commit msg instead of hashes here, less chaotic)
Afterwards the corner-case "this is how RPG_RT does it" stuff begins which needs more looking and testing.

Additional 3 commits also cherry-pickable imo from more down the commit log (if possible conflict free):
 - Remove unused Game_Character::animation_id 
 - Use SaveMapEventBase::anim_frame … 
 - Use SaveMapEventBase::anim_paused 

Only stuff I found: "Encapsulte" in one commit message.
And there are multiple doc comments missing the @param field.

Ghabry commented:
Well, maybe should also give some 👍 . Nice refactor, I found this huge code duplication with virtual inheritance ugly since I added it. xD

And great to see more LSD compat problems solved.

fmatthew5876 commented:
I've rebased this and removed the commits after the requested break point. I've cherry-picked all the ones asked for, including `Utils::Clamp()` as that's needed to make this compile on windows.

Ghabry commented:
Jenkins: test this please  (vcpkg 😘 )

fmatthew5876 commented:
This one starts to become a blocker now. For future LSD work.

carstene1ns commented on src/player.cpp:
```diff
@@ -772,6 +772,15 @@ static void OnMapSaveFileReady(FileRequestResult*) {
 		Main_Data::game_data.party_location.position_y
 		);
+
+	// Compatibility hack for old EasyRPG Player saves.
```
Maybe we should add some version string, e.g. `(pre 0.6)` here for reference.

fmatthew5876 replied:
Great idea. Fixed.
